tensorflow/python/data/kernel_tests/matching_files_dataset_op_test.py

```diff
@@ -0,0 +1,232 @@
+# Copyright 2018 The TensorFlow Authors. All Rights Reserved.
+#
+# Licensed under the Apache License, Version 2.0 (the "License");
+# you may not use this file except in compliance with the License.
+# You may obtain a copy of the License at
+#
+#     http://www.apache.org/licenses/LICENSE-2.0
+#
+# Unless required by applicable law or agreed to in writing, software
+# distributed under the License is distributed on an "AS IS" BASIS,
+# WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+# See the License for the specific language governing permissions and
+# limitations under the License.
+# ==============================================================================
+"""Tests for the experimental input pipeline ops."""
+from __future__ import absolute_import
+from __future__ import division
+from __future__ import print_function
+
+import os
+import shutil
+import tempfile
+import time
+
+import numpy as np
+
+from tensorflow.python.client import session
+from tensorflow.python.data.kernel_tests import test_base
+from tensorflow.python.data.ops.dataset_ops import MatchingFilesDataset
+from tensorflow.python.framework import errors
+from tensorflow.python.framework import ops
+from tensorflow.python.platform import test
+from tensorflow.python.util import compat
+
+
+class MatchingFilesDatasetTest(test_base.DatasetTestBase):
+
+  def setUp(self):
+    self.tmp_dir = tempfile.mkdtemp()
+
+  def tearDown(self):
+    shutil.rmtree(self.tmp_dir, ignore_errors=True)
+
+  def _touchTempFiles(self, filenames):
+    for filename in filenames:
+      open(os.path.join(self.tmp_dir, filename), 'a').close()
+
+  def testNonExistingDirectory(self):
+    """Test the MatchingFiles dataset with a non-existing directory"""
+
+    self.tearDown()
+    dataset = MatchingFilesDataset(os.path.join(self.tmp_dir, '*'))
+    with self.cached_session() as sess:
+      next_element = dataset.make_one_shot_iterator().get_next()
+      with self.assertRaises(errors.NotFoundError):
```

Comment:
this test fails because your implementation does not generate NotFoundError when the pattern does not match anything

Reply:
@jsimsa This test passed on my local laptop(Mac). These [two lines](https://github.com/tensorflow/tensorflow/blob/9be427e21141e1be47cfd30865b7e4c1f434ea86/tensorflow/core/kernels/data/matching_files_dataset_op.cc#L236-L237) will generate a `NotFoundError` if the directory derived from the input pattern does not exist.

Could you share me with the Error log?

Note that in the `test.log` file, there will be some ERROR logs like below. It is caused by [ErrorLoggingSession](https://github.com/tensorflow/tensorflow/blob/master/tensorflow/python/framework/test_util.py#L948-L961), which will catch all the error logs except `OutofRange`. As you can see, at the end of the test.log, the test log says 
```
Ran 7 tests in 0.120s

OK
```  

```
....ERROR:tensorflow:/var/folders/44/dzfy03t93h990mbh3qjy3pjm0000gn/T/tmpkyoyfosh; No such file or directory
	 [[node IteratorGetNext_4 (defined at /anaconda3/lib/python3.6/unittest/case.py:605)  = IteratorGetNext[output_shapes=[[]], output_types=[DT_STRING], _device="/job:localhost/replica:0/task:0/device:CPU:0"](OneShotIterator_4)]]

Caused by op 'IteratorGetNext_4', defined at:
  File "/private/var/tmp/_bazel_fei/5dc7b372a7c45427ff30500d3e22fe26/execroot/org_tensorflow/bazel-out/darwin-opt/bin/tensorflow/python/data/kernel_tests/matching_files_dataset_op_test.runfiles/org_tensorflow/tensorflow/python/data/kernel_tests/matching_files_dataset_op_test.py", line 232, in <module>
    test.main()
  File "/private/var/tmp/_bazel_fei/5dc7b372a7c45427ff30500d3e22fe26/execroot/org_tensorflow/bazel-out/darwin-opt/bin/tensorflow/python/data/kernel_tests/matching_files_dataset_op_test.runfiles/org_tensorflow/tensorflow/python/platform/test.py", line 64, in main
    return _googletest.main(argv)
  File "/private/var/tmp/_bazel_fei/5dc7b372a7c45427ff30500d3e22fe26/execroot/org_tensorflow/bazel-out/darwin-opt/bin/tensorflow/python/data/kernel_tests/matching_files_dataset_op_test.runfiles/org_tensorflow/tensorflow/python/platform/googletest.py", line 100, in main
    benchmark.benchmarks_main(true_main=main_wrapper)
  File "/private/var/tmp/_bazel_fei/5dc7b372a7c45427ff30500d3e22fe26/execroot/org_tensorflow/bazel-out/darwin-opt/bin/tensorflow/python/data/kernel_tests/matching_files_dataset_op_test.runfiles/org_tensorflow/tensorflow/python/platform/benchmark.py", line 344, in benchmarks_main
    true_main()
  File "/private/var/tmp/_bazel_fei/5dc7b372a7c45427ff30500d3e22fe26/execroot/org_tensorflow/bazel-out/darwin-opt/bin/tensorflow/python/data/kernel_tests/matching_files_dataset_op_test.runfiles/org_tensorflow/tensorflow/python/platform/googletest.py", line 99, in main_wrapper
    return app.run(main=g_main, argv=args)
  File "/private/var/tmp/_bazel_fei/5dc7b372a7c45427ff30500d3e22fe26/execroot/org_tensorflow/bazel-out/darwin-opt/bin/tensorflow/python/data/kernel_tests/matching_files_dataset_op_test.runfiles/org_tensorflow/tensorflow/python/platform/app.py", line 125, in run
    _sys.exit(main(argv))
  File "/private/var/tmp/_bazel_fei/5dc7b372a7c45427ff30500d3e22fe26/execroot/org_tensorflow/bazel-out/darwin-opt/bin/tensorflow/python/data/kernel_tests/matching_files_dataset_op_test.runfiles/org_tensorflow/tensorflow/python/platform/googletest.py", line 70, in g_main
    return unittest_main(argv=argv)
  File "/anaconda3/lib/python3.6/unittest/main.py", line 95, in __init__
    self.runTests()
  File "/anaconda3/lib/python3.6/unittest/main.py", line 256, in runTests
    self.result = testRunner.run(self.test)
  File "/anaconda3/lib/python3.6/unittest/runner.py", line 176, in run
    test(result)
  File "/anaconda3/lib/python3.6/unittest/suite.py", line 84, in __call__
    return self.run(*args, **kwds)
  File "/anaconda3/lib/python3.6/unittest/suite.py", line 122, in run
    test(result)
  File "/anaconda3/lib/python3.6/unittest/suite.py", line 84, in __call__
    return self.run(*args, **kwds)
  File "/anaconda3/lib/python3.6/unittest/suite.py", line 122, in run
    test(result)
  File "/anaconda3/lib/python3.6/unittest/case.py", line 653, in __call__
    return self.run(*args, **kwds)
  File "/anaconda3/lib/python3.6/unittest/case.py", line 605, in run
    testMethod()
  File "/private/var/tmp/_bazel_fei/5dc7b372a7c45427ff30500d3e22fe26/execroot/org_tensorflow/bazel-out/darwin-opt/bin/tensorflow/python/data/kernel_tests/matching_files_dataset_op_test.runfiles/org_tensorflow/tensorflow/python/data/kernel_tests/matching_files_dataset_op_test.py", line 54, in testNonExistingDirectory
    next_element = dataset.make_one_shot_iterator().get_next()
  File "/private/var/tmp/_bazel_fei/5dc7b372a7c45427ff30500d3e22fe26/execroot/org_tensorflow/bazel-out/darwin-opt/bin/tensorflow/python/data/kernel_tests/matching_files_dataset_op_test.runfiles/org_tensorflow/tensorflow/python/data/ops/iterator_ops.py", line 421, in get_next
    name=name)), self._output_types,
  File "/private/var/tmp/_bazel_fei/5dc7b372a7c45427ff30500d3e22fe26/execroot/org_tensorflow/bazel-out/darwin-opt/bin/tensorflow/python/data/kernel_tests/matching_files_dataset_op_test.runfiles/org_tensorflow/tensorflow/python/ops/gen_dataset_ops.py", line 2069, in iterator_get_next
    output_shapes=output_shapes, name=name)
  File "/private/var/tmp/_bazel_fei/5dc7b372a7c45427ff30500d3e22fe26/execroot/org_tensorflow/bazel-out/darwin-opt/bin/tensorflow/python/data/kernel_tests/matching_files_dataset_op_test.runfiles/org_tensorflow/tensorflow/python/framework/op_def_library.py", line 787, in _apply_op_helper
    op_def=op_def)
  File "/private/var/tmp/_bazel_fei/5dc7b372a7c45427ff30500d3e22fe26/execroot/org_tensorflow/bazel-out/darwin-opt/bin/tensorflow/python/data/kernel_tests/matching_files_dataset_op_test.runfiles/org_tensorflow/tensorflow/python/util/deprecation.py", line 488, in new_func
    return func(*args, **kwargs)
  File "/private/var/tmp/_bazel_fei/5dc7b372a7c45427ff30500d3e22fe26/execroot/org_tensorflow/bazel-out/darwin-opt/bin/tensorflow/python/data/kernel_tests/matching_files_dataset_op_test.runfiles/org_tensorflow/tensorflow/python/framework/ops.py", line 3274, in create_op
    op_def=op_def)
  File "/private/var/tmp/_bazel_fei/5dc7b372a7c45427ff30500d3e22fe26/execroot/org_tensorflow/bazel-out/darwin-opt/bin/tensorflow/python/data/kernel_tests/matching_files_dataset_op_test.runfiles/org_tensorflow/tensorflow/python/framework/ops.py", line 1770, in __init__
    self._traceback = tf_stack.extract_stack()

NotFoundError (see above for traceback): /var/folders/44/dzfy03t93h990mbh3qjy3pjm0000gn/T/tmpkyoyfosh; No such file or directory
	 [[node IteratorGetNext_4 (defined at /anaconda3/lib/python3.6/unittest/case.py:605)  = IteratorGetNext[output_shapes=[[]], output_types=[DT_STRING], _device="/job:localhost/replica:0/task:0/device:CPU:0"](OneShotIterator_4)]]

...
----------------------------------------------------------------------
Ran 7 tests in 0.120s

OK
```

Reply:
```
==================== Test output for //py_test_dir/tensorflow/python/data/kernel_tests:matching_files_dataset_op_test:
WARNING:tensorflow:From C:\Python36\lib\site-packages\tensorflow\python\ops\distributions\distribution.py:266: ReparameterizationType.__init__ (from tensorflow.python.ops.distributions.distribution) is deprecated and will be removed after 2019-01-01.
Instructions for updating:
The TensorFlow Distributions library has moved to TensorFlow Probability (https://github.com/tensorflow/probability). You should update all references to use `tfp.distributions` instead of `tf.distributions`.
WARNING:tensorflow:From C:\Python36\lib\site-packages\tensorflow\python\ops\distributions\bernoulli.py:169: RegisterKL.__init__ (from tensorflow.python.ops.distributions.kullback_leibler) is deprecated and will be removed after 2019-01-01.
Instructions for updating:
The TensorFlow Distributions library has moved to TensorFlow Probability (https://github.com/tensorflow/probability). You should update all references to use `tfp.distributions` instead of `tf.distributions`.
2018-10-05 18:29:05.255826: I tensorflow/core/platform/cpu_feature_guard.cc:141] Your CPU supports instructions that this TensorFlow binary was not compiled to use: AVX2
.EEFEE.
======================================================================
ERROR: testFileMiddles (__main__.MatchingFilesDatasetTest)
Test the MatchingFiles dataset using the middles of filename
----------------------------------------------------------------------
Traceback (most recent call last):
  File "C:\Python36\lib\site-packages\tensorflow\python\client\session.py", line 1335, in _do_call
    return fn(*args)
  File "C:\Python36\lib\site-packages\tensorflow\python\client\session.py", line 1320, in _run_fn
    options, feed_dict, fetch_list, target_list, run_metadata)
  File "C:\Python36\lib\site-packages\tensorflow\python\client\session.py", line 1408, in _call_tf_sessionrun
    run_metadata)
tensorflow.python.framework.errors_impl.OutOfRangeError: End of sequence
	 [[{{node IteratorGetNext_1}} = IteratorGetNext[output_shapes=[[]], output_types=[DT_STRING], _device="/job:localhost/replica:0/task:0/device:CPU:0"](OneShotIterator_1)]]

During handling of the above exception, another exception occurred:

Traceback (most recent call last):
  File "\\?\T:\tmp\Bazel.runfiles_p48sp2bq\runfiles\org_tensorflow\py_test_dir\tensorflow\python\data\kernel_tests\matching_files_dataset_op_test.py", line 122, in testFileMiddles
    actual_filenames.append(compat.as_bytes(sess.run(next_element)))
  File "C:\Python36\lib\site-packages\tensorflow\python\framework\test_util.py", line 955, in run
    return super(ErrorLoggingSession, self).run(*args, **kwargs)
  File "C:\Python36\lib\site-packages\tensorflow\python\client\session.py", line 930, in run
    run_metadata_ptr)
  File "C:\Python36\lib\site-packages\tensorflow\python\client\session.py", line 1153, in _run
    feed_dict_tensor, options, run_metadata)
  File "C:\Python36\lib\site-packages\tensorflow\python\client\session.py", line 1329, in _do_run
    run_metadata)
  File "C:\Python36\lib\site-packages\tensorflow\python\client\session.py", line 1349, in _do_call
    raise type(e)(node_def, op, message)
tensorflow.python.framework.errors_impl.OutOfRangeError: End of sequence
	 [[node IteratorGetNext_1 (defined at C:\Python36\lib\unittest\case.py:605)  = IteratorGetNext[output_shapes=[[]], output_types=[DT_STRING], _device="/job:localhost/replica:0/task:0/device:CPU:0"](OneShotIterator_1)]]

Caused by op 'IteratorGetNext_1', defined at:
  File "\\?\T:\tmp\Bazel.runfiles_p48sp2bq\runfiles\org_tensorflow\py_test_dir\tensorflow\python\data\kernel_tests\matching_files_dataset_op_test.py", line 236, in <module>
    test.main()
  File "C:\Python36\lib\site-packages\tensorflow\python\platform\test.py", line 64, in main
    return _googletest.main(argv)
  File "C:\Python36\lib\site-packages\tensorflow\python\platform\googletest.py", line 100, in main
    benchmark.benchmarks_main(true_main=main_wrapper)
  File "C:\Python36\lib\site-packages\tensorflow\python\platform\benchmark.py", line 345, in benchmarks_main
    true_main()
  File "C:\Python36\lib\site-packages\tensorflow\python\platform\googletest.py", line 99, in main_wrapper
    return app.run(main=g_main, argv=args)
  File "C:\Python36\lib\site-packages\tensorflow\python\platform\app.py", line 125, in run
    _sys.exit(main(argv))
  File "C:\Python36\lib\site-packages\tensorflow\python\platform\googletest.py", line 70, in g_main
    return unittest_main(argv=argv)
  File "C:\Python36\lib\unittest\main.py", line 95, in __init__
    self.runTests()
  File "C:\Python36\lib\unittest\main.py", line 256, in runTests
    self.result = testRunner.run(self.test)
  File "C:\Python36\lib\unittest\runner.py", line 176, in run
    test(result)
  File "C:\Python36\lib\unittest\suite.py", line 84, in __call__
    return self.run(*args, **kwds)
  File "C:\Python36\lib\unittest\suite.py", line 122, in run
    test(result)
  File "C:\Python36\lib\unittest\suite.py", line 84, in __call__
    return self.run(*args, **kwds)
  File "C:\Python36\lib\unittest\suite.py", line 122, in run
    test(result)
  File "C:\Python36\lib\unittest\case.py", line 653, in __call__
    return self.run(*args, **kwds)
  File "C:\Python36\lib\unittest\case.py", line 605, in run
    testMethod()
  File "\\?\T:\tmp\Bazel.runfiles_p48sp2bq\runfiles\org_tensorflow\py_test_dir\tensorflow\python\data\kernel_tests\matching_files_dataset_op_test.py", line 116, in testFileMiddles
    next_element = dataset.make_one_shot_iterator().get_next()
  File "C:\Python36\lib\site-packages\tensorflow\python\data\ops\iterator_ops.py", line 421, in get_next
    name=name)), self._output_types,
  File "C:\Python36\lib\site-packages\tensorflow\python\ops\gen_dataset_ops.py", line 2107, in iterator_get_next
    output_shapes=output_shapes, name=name)
  File "C:\Python36\lib\site-packages\tensorflow\python\framework\op_def_library.py", line 788, in _apply_op_helper
    op_def=op_def)
  File "C:\Python36\lib\site-packages\tensorflow\python\util\deprecation.py", line 488, in new_func
    return func(*args, **kwargs)
  File "C:\Python36\lib\site-packages\tensorflow\python\framework\ops.py", line 3275, in create_op
    op_def=op_def)
  File "C:\Python36\lib\site-packages\tensorflow\python\framework\ops.py", line 1771, in __init__
    self._traceback = tf_stack.extract_stack()

OutOfRangeError (see above for traceback): End of sequence
	 [[node IteratorGetNext_1 (defined at C:\Python36\lib\unittest\case.py:605)  = IteratorGetNext[output_shapes=[[]], output_types=[DT_STRING], _device="/job:localhost/replica:0/task:0/device:CPU:0"](OneShotIterator_1)]]


======================================================================
ERROR: testFileSuffixes (__main__.MatchingFilesDatasetTest)
Test the MatchingFiles dataset using the suffixes of filename
----------------------------------------------------------------------
Traceback (most recent call last):
  File "C:\Python36\lib\site-packages\tensorflow\python\client\session.py", line 1335, in _do_call
    return fn(*args)
  File "C:\Python36\lib\site-packages\tensorflow\python\client\session.py", line 1320, in _run_fn
    options, feed_dict, fetch_list, target_list, run_metadata)
  File "C:\Python36\lib\site-packages\tensorflow\python\client\session.py", line 1408, in _call_tf_sessionrun
    run_metadata)
tensorflow.python.framework.errors_impl.OutOfRangeError: End of sequence
	 [[{{node IteratorGetNext_2}} = IteratorGetNext[output_shapes=[[]], output_types=[DT_STRING], _device="/job:localhost/replica:0/task:0/device:CPU:0"](OneShotIterator_2)]]

During handling of the above exception, another exception occurred:

Traceback (most recent call last):
  File "\\?\T:\tmp\Bazel.runfiles_p48sp2bq\runfiles\org_tensorflow\py_test_dir\tensorflow\python\data\kernel_tests\matching_files_dataset_op_test.py", line 102, in testFileSuffixes
    actual_filenames.append(compat.as_bytes(sess.run(next_element)))
  File "C:\Python36\lib\site-packages\tensorflow\python\framework\test_util.py", line 955, in run
    return super(ErrorLoggingSession, self).run(*args, **kwargs)
  File "C:\Python36\lib\site-packages\tensorflow\python\client\session.py", line 930, in run
    run_metadata_ptr)
  File "C:\Python36\lib\site-packages\tensorflow\python\client\session.py", line 1153, in _run
    feed_dict_tensor, options, run_metadata)
  File "C:\Python36\lib\site-packages\tensorflow\python\client\session.py", line 1329, in _do_run
    run_metadata)
  File "C:\Python36\lib\site-packages\tensorflow\python\client\session.py", line 1349, in _do_call
    raise type(e)(node_def, op, message)
tensorflow.python.framework.errors_impl.OutOfRangeError: End of sequence
	 [[node IteratorGetNext_2 (defined at C:\Python36\lib\unittest\case.py:605)  = IteratorGetNext[output_shapes=[[]], output_types=[DT_STRING], _device="/job:localhost/replica:0/task:0/device:CPU:0"](OneShotIterator_2)]]

Caused by op 'IteratorGetNext_2', defined at:
  File "\\?\T:\tmp\Bazel.runfiles_p48sp2bq\runfiles\org_tensorflow\py_test_dir\tensorflow\python\data\kernel_tests\matching_files_dataset_op_test.py", line 236, in <module>
    test.main()
  File "C:\Python36\lib\site-packages\tensorflow\python\platform\test.py", line 64, in main
    return _googletest.main(argv)
  File "C:\Python36\lib\site-packages\tensorflow\python\platform\googletest.py", line 100, in main
    benchmark.benchmarks_main(true_main=main_wrapper)
  File "C:\Python36\lib\site-packages\tensorflow\python\platform\benchmark.py", line 345, in benchmarks_main
    true_main()
  File "C:\Python36\lib\site-packages\tensorflow\python\platform\googletest.py", line 99, in main_wrapper
    return app.run(main=g_main, argv=args)
  File "C:\Python36\lib\site-packages\tensorflow\python\platform\app.py", line 125, in run
    _sys.exit(main(argv))
  File "C:\Python36\lib\site-packages\tensorflow\python\platform\googletest.py", line 70, in g_main
    return unittest_main(argv=argv)
  File "C:\Python36\lib\unittest\main.py", line 95, in __init__
    self.runTests()
  File "C:\Python36\lib\unittest\main.py", line 256, in runTests
    self.result = testRunner.run(self.test)
  File "C:\Python36\lib\unittest\runner.py", line 176, in run
    test(result)
  File "C:\Python36\lib\unittest\suite.py", line 84, in __call__
    return self.run(*args, **kwds)
  File "C:\Python36\lib\unittest\suite.py", line 122, in run
    test(result)
  File "C:\Python36\lib\unittest\suite.py", line 84, in __call__
    return self.run(*args, **kwds)
  File "C:\Python36\lib\unittest\suite.py", line 122, in run
    test(result)
  File "C:\Python36\lib\unittest\case.py", line 653, in __call__
    return self.run(*args, **kwds)
  File "C:\Python36\lib\unittest\case.py", line 605, in run
    testMethod()
  File "\\?\T:\tmp\Bazel.runfiles_p48sp2bq\runfiles\org_tensorflow\py_test_dir\tensorflow\python\data\kernel_tests\matching_files_dataset_op_test.py", line 96, in testFileSuffixes
    next_element = dataset.make_one_shot_iterator().get_next()
  File "C:\Python36\lib\site-packages\tensorflow\python\data\ops\iterator_ops.py", line 421, in get_next
    name=name)), self._output_types,
  File "C:\Python36\lib\site-packages\tensorflow\python\ops\gen_dataset_ops.py", line 2107, in iterator_get_next
    output_shapes=output_shapes, name=name)
  File "C:\Python36\lib\site-packages\tensorflow\python\framework\op_def_library.py", line 788, in _apply_op_helper
    op_def=op_def)
  File "C:\Python36\lib\site-packages\tensorflow\python\util\deprecation.py", line 488, in new_func
    return func(*args, **kwargs)
  File "C:\Python36\lib\site-packages\tensorflow\python\framework\ops.py", line 3275, in create_op
    op_def=op_def)
  File "C:\Python36\lib\site-packages\tensorflow\python\framework\ops.py", line 1771, in __init__
    self._traceback = tf_stack.extract_stack()

OutOfRangeError (see above for traceback): End of sequence
	 [[node IteratorGetNext_2 (defined at C:\Python36\lib\unittest\case.py:605)  = IteratorGetNext[output_shapes=[[]], output_types=[DT_STRING], _device="/job:localhost/replica:0/task:0/device:CPU:0"](OneShotIterator_2)]]


======================================================================
ERROR: testNonExistingDirectory (__main__.MatchingFilesDatasetTest)
Test the MatchingFiles dataset with a non-existing directory
----------------------------------------------------------------------
Traceback (most recent call last):
  File "C:\Python36\lib\site-packages\tensorflow\python\client\session.py", line 1335, in _do_call
    return fn(*args)
  File "C:\Python36\lib\site-packages\tensorflow\python\client\session.py", line 1320, in _run_fn
    options, feed_dict, fetch_list, target_list, run_metadata)
  File "C:\Python36\lib\site-packages\tensorflow\python\client\session.py", line 1408, in _call_tf_sessionrun
    run_metadata)
tensorflow.python.framework.errors_impl.OutOfRangeError: End of sequence
	 [[{{node IteratorGetNext_4}} = IteratorGetNext[output_shapes=[[]], output_types=[DT_STRING], _device="/job:localhost/replica:0/task:0/device:CPU:0"](OneShotIterator_4)]]

During handling of the above exception, another exception occurred:

Traceback (most recent call last):
  File "\\?\T:\tmp\Bazel.runfiles_p48sp2bq\runfiles\org_tensorflow\py_test_dir\tensorflow\python\data\kernel_tests\matching_files_dataset_op_test.py", line 56, in testNonExistingDirectory
    sess.run(next_element)
  File "C:\Python36\lib\site-packages\tensorflow\python\framework\test_util.py", line 955, in run
    return super(ErrorLoggingSession, self).run(*args, **kwargs)
  File "C:\Python36\lib\site-packages\tensorflow\python\client\session.py", line 930, in run
    run_metadata_ptr)
  File "C:\Python36\lib\site-packages\tensorflow\python\client\session.py", line 1153, in _run
    feed_dict_tensor, options, run_metadata)
  File "C:\Python36\lib\site-packages\tensorflow\python\client\session.py", line 1329, in _do_run
    run_metadata)
  File "C:\Python36\lib\site-packages\tensorflow\python\client\session.py", line 1349, in _do_call
    raise type(e)(node_def, op, message)
tensorflow.python.framework.errors_impl.OutOfRangeError: End of sequence
	 [[node IteratorGetNext_4 (defined at C:\Python36\lib\unittest\case.py:605)  = IteratorGetNext[output_shapes=[[]], output_types=[DT_STRING], _device="/job:localhost/replica:0/task:0/device:CPU:0"](OneShotIterator_4)]]

Caused by op 'IteratorGetNext_4', defined at:
  File "\\?\T:\tmp\Bazel.runfiles_p48sp2bq\runfiles\org_tensorflow\py_test_dir\tensorflow\python\data\kernel_tests\matching_files_dataset_op_test.py", line 236, in <module>
    test.main()
  File "C:\Python36\lib\site-packages\tensorflow\python\platform\test.py", line 64, in main
    return _googletest.main(argv)
  File "C:\Python36\lib\site-packages\tensorflow\python\platform\googletest.py", line 100, in main
    benchmark.benchmarks_main(true_main=main_wrapper)
  File "C:\Python36\lib\site-packages\tensorflow\python\platform\benchmark.py", line 345, in benchmarks_main
    true_main()
  File "C:\Python36\lib\site-packages\tensorflow\python\platform\googletest.py", line 99, in main_wrapper
    return app.run(main=g_main, argv=args)
  File "C:\Python36\lib\site-packages\tensorflow\python\platform\app.py", line 125, in run
    _sys.exit(main(argv))
  File "C:\Python36\lib\site-packages\tensorflow\python\platform\googletest.py", line 70, in g_main
    return unittest_main(argv=argv)
  File "C:\Python36\lib\unittest\main.py", line 95, in __init__
    self.runTests()
  File "C:\Python36\lib\unittest\main.py", line 256, in runTests
    self.result = testRunner.run(self.test)
  File "C:\Python36\lib\unittest\runner.py", line 176, in run
    test(result)
  File "C:\Python36\lib\unittest\suite.py", line 84, in __call__
    return self.run(*args, **kwds)
  File "C:\Python36\lib\unittest\suite.py", line 122, in run
    test(result)
  File "C:\Python36\lib\unittest\suite.py", line 84, in __call__
    return self.run(*args, **kwds)
  File "C:\Python36\lib\unittest\suite.py", line 122, in run
    test(result)
  File "C:\Python36\lib\unittest\case.py", line 653, in __call__
    return self.run(*args, **kwds)
  File "C:\Python36\lib\unittest\case.py", line 605, in run
    testMethod()
  File "\\?\T:\tmp\Bazel.runfiles_p48sp2bq\runfiles\org_tensorflow\py_test_dir\tensorflow\python\data\kernel_tests\matching_files_dataset_op_test.py", line 54, in testNonExistingDirectory
    next_element = dataset.make_one_shot_iterator().get_next()
  File "C:\Python36\lib\site-packages\tensorflow\python\data\ops\iterator_ops.py", line 421, in get_next
    name=name)), self._output_types,
  File "C:\Python36\lib\site-packages\tensorflow\python\ops\gen_dataset_ops.py", line 2107, in iterator_get_next
    output_shapes=output_shapes, name=name)
  File "C:\Python36\lib\site-packages\tensorflow\python\framework\op_def_library.py", line 788, in _apply_op_helper
    op_def=op_def)
  File "C:\Python36\lib\site-packages\tensorflow\python\util\deprecation.py", line 488, in new_func
    return func(*args, **kwargs)
  File "C:\Python36\lib\site-packages\tensorflow\python\framework\ops.py", line 3275, in create_op
    op_def=op_def)
  File "C:\Python36\lib\site-packages\tensorflow\python\framework\ops.py", line 1771, in __init__
    self._traceback = tf_stack.extract_stack()

OutOfRangeError (see above for traceback): End of sequence
	 [[node IteratorGetNext_4 (defined at C:\Python36\lib\unittest\case.py:605)  = IteratorGetNext[output_shapes=[[]], output_types=[DT_STRING], _device="/job:localhost/replica:0/task:0/device:CPU:0"](OneShotIterator_4)]]


======================================================================
ERROR: testSimpleDirectory (__main__.MatchingFilesDatasetTest)
Test the MatchingFiles dataset with a simple directory
----------------------------------------------------------------------
Traceback (most recent call last):
  File "C:\Python36\lib\site-packages\tensorflow\python\client\session.py", line 1335, in _do_call
    return fn(*args)
  File "C:\Python36\lib\site-packages\tensorflow\python\client\session.py", line 1320, in _run_fn
    options, feed_dict, fetch_list, target_list, run_metadata)
  File "C:\Python36\lib\site-packages\tensorflow\python\client\session.py", line 1408, in _call_tf_sessionrun
    run_metadata)
tensorflow.python.framework.errors_impl.OutOfRangeError: End of sequence
	 [[{{node IteratorGetNext_5}} = IteratorGetNext[output_shapes=[[]], output_types=[DT_STRING], _device="/job:localhost/replica:0/task:0/device:CPU:0"](OneShotIterator_5)]]

During handling of the above exception, another exception occurred:

Traceback (most recent call last):
  File "\\?\T:\tmp\Bazel.runfiles_p48sp2bq\runfiles\org_tensorflow\py_test_dir\tensorflow\python\data\kernel_tests\matching_files_dataset_op_test.py", line 82, in testSimpleDirectory
    actual_filenames.append(compat.as_bytes(sess.run(next_element)))
  File "C:\Python36\lib\site-packages\tensorflow\python\framework\test_util.py", line 955, in run
    return super(ErrorLoggingSession, self).run(*args, **kwargs)
  File "C:\Python36\lib\site-packages\tensorflow\python\client\session.py", line 930, in run
    run_metadata_ptr)
  File "C:\Python36\lib\site-packages\tensorflow\python\client\session.py", line 1153, in _run
    feed_dict_tensor, options, run_metadata)
  File "C:\Python36\lib\site-packages\tensorflow\python\client\session.py", line 1329, in _do_run
    run_metadata)
  File "C:\Python36\lib\site-packages\tensorflow\python\client\session.py", line 1349, in _do_call
    raise type(e)(node_def, op, message)
tensorflow.python.framework.errors_impl.OutOfRangeError: End of sequence
	 [[node IteratorGetNext_5 (defined at C:\Python36\lib\unittest\case.py:605)  = IteratorGetNext[output_shapes=[[]], output_types=[DT_STRING], _device="/job:localhost/replica:0/task:0/device:CPU:0"](OneShotIterator_5)]]

Caused by op 'IteratorGetNext_5', defined at:
  File "\\?\T:\tmp\Bazel.runfiles_p48sp2bq\runfiles\org_tensorflow\py_test_dir\tensorflow\python\data\kernel_tests\matching_files_dataset_op_test.py", line 236, in <module>
    test.main()
  File "C:\Python36\lib\site-packages\tensorflow\python\platform\test.py", line 64, in main
    return _googletest.main(argv)
  File "C:\Python36\lib\site-packages\tensorflow\python\platform\googletest.py", line 100, in main
    benchmark.benchmarks_main(true_main=main_wrapper)
  File "C:\Python36\lib\site-packages\tensorflow\python\platform\benchmark.py", line 345, in benchmarks_main
    true_main()
  File "C:\Python36\lib\site-packages\tensorflow\python\platform\googletest.py", line 99, in main_wrapper
    return app.run(main=g_main, argv=args)
  File "C:\Python36\lib\site-packages\tensorflow\python\platform\app.py", line 125, in run
    _sys.exit(main(argv))
  File "C:\Python36\lib\site-packages\tensorflow\python\platform\googletest.py", line 70, in g_main
    return unittest_main(argv=argv)
  File "C:\Python36\lib\unittest\main.py", line 95, in __init__
    self.runTests()
  File "C:\Python36\lib\unittest\main.py", line 256, in runTests
    self.result = testRunner.run(self.test)
  File "C:\Python36\lib\unittest\runner.py", line 176, in run
    test(result)
  File "C:\Python36\lib\unittest\suite.py", line 84, in __call__
    return self.run(*args, **kwds)
  File "C:\Python36\lib\unittest\suite.py", line 122, in run
    test(result)
  File "C:\Python36\lib\unittest\suite.py", line 84, in __call__
    return self.run(*args, **kwds)
  File "C:\Python36\lib\unittest\suite.py", line 122, in run
    test(result)
  File "C:\Python36\lib\unittest\case.py", line 653, in __call__
    return self.run(*args, **kwds)
  File "C:\Python36\lib\unittest\case.py", line 605, in run
    testMethod()
  File "\\?\T:\tmp\Bazel.runfiles_p48sp2bq\runfiles\org_tensorflow\py_test_dir\tensorflow\python\data\kernel_tests\matching_files_dataset_op_test.py", line 75, in testSimpleDirectory
    next_element = dataset.make_one_shot_iterator().get_next()
  File "C:\Python36\lib\site-packages\tensorflow\python\data\ops\iterator_ops.py", line 421, in get_next
    name=name)), self._output_types,
  File "C:\Python36\lib\site-packages\tensorflow\python\ops\gen_dataset_ops.py", line 2107, in iterator_get_next
    output_shapes=output_shapes, name=name)
  File "C:\Python36\lib\site-packages\tensorflow\python\framework\op_def_library.py", line 788, in _apply_op_helper
    op_def=op_def)
  File "C:\Python36\lib\site-packages\tensorflow\python\util\deprecation.py", line 488, in new_func
    return func(*args, **kwargs)
  File "C:\Python36\lib\site-packages\tensorflow\python\framework\ops.py", line 3275, in create_op
    op_def=op_def)
  File "C:\Python36\lib\site-packages\tensorflow\python\framework\ops.py", line 1771, in __init__
    self._traceback = tf_stack.extract_stack()

OutOfRangeError (see above for traceback): End of sequence
	 [[node IteratorGetNext_5 (defined at C:\Python36\lib\unittest\case.py:605)  = IteratorGetNext[output_shapes=[[]], output_types=[DT_STRING], _device="/job:localhost/replica:0/task:0/device:CPU:0"](OneShotIterator_5)]]


======================================================================
FAIL: testNestedDirectories (__main__.MatchingFilesDatasetTest)
Test the MatchingFiles dataset with nested directories
----------------------------------------------------------------------
Traceback (most recent call last):
  File "\\?\T:\tmp\Bazel.runfiles_p48sp2bq\runfiles\org_tensorflow\py_test_dir\tensorflow\python\data\kernel_tests\matching_files_dataset_op_test.py", line 166, in testNestedDirectories
    self.assertItemsEqual(expected_filenames, actual_filenames)
AssertionError: Element counts were not equal:

Diff is 2127 characters long. Set self.maxDiff to None to see it.

----------------------------------------------------------------------
Ran 7 tests in 0.379s
```

Reply:
@jsimsa It looks like all the tests fail on Windows? It is not just about `NotFoundError`. I did not have a Window environment to reproduce these errors. Do you have any suggestion?

Reply:
Windows testing aside, my best guess about what could be going wrong is that the error handling logic of `GetNextInternal` is incorrect. Why do we not return as soon as `ret` is updated? In other words, would it make sense to replace `ret.Update(UpdateIterator(...))` with `TF_RETURN_IF_ERROR(UpdateIterator(...))`?

Reply:
@jsimsa Your suggestion makes sense. In the beginning, I thought a scenario that user inputs two patterns (the first pattern is invalid (e.g. the directory doesn't exist), and the second is valid). With `ret.Update(UpdateIterator(...))`, it will continue searching the second pattern to return the matched files. With `TF_RETURN_IF_ERROR(UpdateIterator(...)) `, it will return the error immediately without searching the second valid pattern. Yeah, `TF_RETURN_IF_ERROR(UpdateIterator(...)) ` is better. I will change it.

What do you think about the windows testing?

Reply:
If that's the case, I would not make the change I suggested because I think that we should return `NotFoundError` iff there is no match for any of the patterns.

Looking at your code, I we currently return non-ok status when either `env()->GetFileSystemForFile` fails (line 216) or `fs->GetChildren` fails (line 236).

I suggest you add temporary log statements that tells us what the returned status at these lines is which will likely help us identify the root cause of the failures.

I think that you will ultimately need to treat `NotFoundError` differently from other non-ok statuses. For all non-ok status other that `NotFoundError` we should propagate this to the caller right away (as opposed to continue doing other work which can also result in an error that will be discarded). For `NotFoundError`, we should skip over these and only return `NotFoundError` to the caller if all patterns evaluated to an empty set of files.